### DISCUSSION TITLE
Docker networks for AIO

### DISCUSSION THREAD

**Discussion by loeffelpan** [nextcloud/all-in-one#1630] [Questions] [answered]
AIO is up and running.
Docker shows three network with aio-containers connected.
Most aio-containers are connected with `nextcloud-aio`-network AND `bridge`-network.
The mastercontainer is actually connected to `nextcloud-aio`-, `bridge`- and `nextcloud_default`-network.

Is that intended?
With this configuration there are many `veth`-adapters.

**Comment by szaimen** [comment]
> Most aio-containers are connected with `nextcloud-aio`-network AND `bridge`-network.

Yes, this is normal. However we only connect it to the nextcloud-aio network which is of kind bridge.
See https://github.com/nextcloud/all-in-one/blob/9a9f16ec2fb822690d244b08354876b1edf2abe5/php/src/Docker/DockerActionManager.php#L618-L624


> The mastercontainer is actually connected to `nextcloud-aio`-, `bridge`- and `nextcloud_default`-network.

The first two are normal as stated above. The nextcloud_default might get added automatically due to docker-compose if you have not added it manually? I don't know...

**Reply by loeffelpan** [reply]
Thanks for your reply.
I know, that the `nextcloud-aio` is a brigde network.
But my containers are connected to the default docker network called `bridge`. See below.
Just wondering if this is intended.

So `nextcloud_default` seem to come with docker-compose. Can I remove it?

```
sudo docker network ls
NETWORK ID     NAME                DRIVER    SCOPE
18bda7670acf   bridge              bridge    local
ee616c407e8b   diun_default        bridge    local
60902b46c9b1   host                host      local
12db0838da25   nextcloud-aio       bridge    local
93bdd8f2ba37   nextcloud_default   bridge    local
65cffef058a5   none                null      local

sudo docker network inspect bridge
[...]
        "Containers": {
            "1c7b60ef9195f790def1dd678d90d70ec5bdca387883312028f51e1f5f79588e": {
                "Name": "nextcloud-aio-database",
                "EndpointID": "ea45c59e34e0cdda7ffcf749e292d22266a87107c40976e36df9cd3d80e12385",
                "MacAddress": "02:42:ac:14:00:03",
                "IPv4Address": "172.20.0.3/24",
                "IPv6Address": ""
            },
            "3e6b0deb4f1c0a1aaa2f2eeed6aae6becd921332717706e5d9f1d583544ac77e": {
                "Name": "nextcloud-aio-imaginary",
                "EndpointID": "d874691fa426125c25c07749326ab4246d49c70701487ea53e8a29de17bbc033",
                "MacAddress": "02:42:ac:14:00:05",
                "IPv4Address": "172.20.0.5/24",
                "IPv6Address": ""
            },
            "5ad550c708b42e1a517c506d833f8859e4d9b5b0e9a2379437d6a1e8bcac2975": {
                "Name": "nextcloud-aio-redis",
                "EndpointID": "9e74216181ce5de46d6bdfc774596339f3229133d65f316bd1435ca39dc415bb",
                "MacAddress": "02:42:ac:14:00:04",
                "IPv4Address": "172.20.0.4/24",
                "IPv6Address": ""
            },
            "6475fff24f786dcb00d26f2c7b7542bf7b06c3ff68d1224783de236ada6f793d": {
                "Name": "nextcloud-aio-collabora",
                "EndpointID": "36f4e8ac11d77ba87f611a7e05375b7bb6ab5cc45803b1e2e330cbd46edb319f",
                "MacAddress": "02:42:ac:14:00:02",
                "IPv4Address": "172.20.0.2/24",
                "IPv6Address": ""
            },
            "71300e8e683b0ac5c14c4dbec72a6ec224a4e80f9f0952fa432dbb43eba2c7ec": {
                "Name": "nextcloud-aio-nextcloud",
                "EndpointID": "210f09da1a40a2e0b2d3202ba5c9635a51f49c526be5d388e1ea5ca1fdbc2d0e",
                "MacAddress": "02:42:ac:14:00:06",
                "IPv4Address": "172.20.0.6/24",
                "IPv6Address": ""
            },
            "e551e3ea778580d46a76182f68668c6bcd8c6785ce79a2c2fceda49b6e1ef3ef": {
                "Name": "nextcloud-aio-apache",
                "EndpointID": "0e7714510ef2897c64f291368e1f6642f9d75da887c2b48d424387eefca49590",
                "MacAddress": "02:42:ac:14:00:07",
                "IPv4Address": "172.20.0.7/24",
                "IPv6Address": ""
            }
        },
[...]
```

**Comment by szaimen** [comment, marked as answer]
> But my containers are connected to the default docker network called `bride`. See below.
> Just wondering if this is intended.

it is not intended but no idea how to resolve that since we don't add it to the bridge network manuallly. Also since it causes no harm being connected to the bridge network I have not investigated further.

> So nextcloud_default seem to come with docker-compose. Can I remove it?

Yes you can remove it but I think that docker-compose will likely recreate it automatically.

**Reply by loeffelpan** [reply]
Thanks anyway.
Just causing six more (useless) `veth...` interfaces in linux.

**Reply by szaimen** [reply]
This is now fixed with v4.4.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Reply by loeffelpan** [reply]
Both of them?
I would guess the `nextcloud_default` one is compose default behavior and will persist. Correct?
I will give it a test during the weekend.

**Reply by szaimen** [reply]
> I would guess the `nextcloud_default` one is compose default behavior and will persist. Correct?

Yes

**Reply by loeffelpan** [reply]
Working for me.
I worked around `nextcloud_default` with setting up a custom bridge named `nextcloud-aio` in my compose file. So the mastercontainer is attached directly to that bridge you would attach it at runtime.

**Reply by Zoey2936** [reply]
I'm not sure, if this is such a good idea, because what, if we at some point for whatever reason need to set custom changes to the nextcloud-aio network, this could break your entire setup

**Reply by loeffelpan** [reply]
OK, got your point.
I tried to add `network_mode: bridge` to my compose file, to add mastercontainer to the default bridge like `docker run` would do.
But I noticed that after starting you remove it from default `bridge`-network.
If you want to have mastercontainer on another network than `nextcloud-aio` you should leave it alone. As mentoined above compose would add a `<foldername>_default`-network but `docker run` adds it to default `bridge`-network and current beta removes it from there.

**Reply by Zoey2936** [reply]
seems to work, do you might create a pull request?

**Reply by Zoey2936** [reply]
@szaimen what do you think? If `docker run` adds the container by default to the bridge network, we could also add the mastercontainer inside the compose file to the bridge network.

**Reply by loeffelpan** [reply]
And keep it whereever `compose` or `run` have attached it, if it's not a good idea to have `mastercontainer` only on your custom bridge. Current beta seems to remove `mastercontainer` from `bridge` like the other containers.

**Reply by Zoey2936** [reply]
> And keep it whereever `compose` or `run` have attached it, if it's not a good idea to have `mastercontainer` only on your custom bridge. Current beta seems to remove `mastercontainer` from `bridge` like the other containers.

I see there no reason to have the mastercontainer inside two networks

**Reply by loeffelpan** [reply]
> I'm not sure, if this is such a good idea, because what, if we at some point for whatever reason need to set custom changes to the nextcloud-aio network, this could break your entire setup

So this i only because of setting up the custom bridge by myself? Thought there are connectivity reasons to reach the mastercontainer when custom bridge is recreated for some reason.


If not, adding it to the default bridge initially and then attach it to the custom bridge would be the best way.

**Reply by Zoey2936** [reply]
The mastercontainer attaches all container to the nextcloud-aio network and removes all container from the default bridge network, and because we may need to make custom changes to the nextcloud-aio for a reason we don't know at the moment, but it could happen, your own created network, wouldn't have this changes, which could break AIO. And for the mastercontainer, It's enough to have it inside the nextcloud-aio network, so adding it to the bridge network, from where it removes itself seems not be a problem, since it's enough to have it inside the nextcloud-aio network.
